### PR TITLE
<DatePicker/> - Adding Unidriver

### PR DESCRIPTION
**Pull request by roibWix**
### 🔦 Summary

...

### ✅ Checklist

- [ ] 👨‍💻 API change is approved by the UI Infra Developers 
- [ ] 👨‍🎨 UX change is approved by the Design System UX 
- 📚 Change is documented
  - [ ] Story
  - [ ] API description
  - [ ] Cheatsheet
  - [ ] Other (explain)
- 🔬 Change is tested
  - [ ] Component
  - [ ] Visual test

### CODE REVIEW

**Comment by roibWix** [comment]
@EgoziE 
I'm pretty sure it's the wrong way to do it in the Input.uni.driver.js.
Let me know what you think.

**Comment by WixBuildServer** [comment]
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://preview.wix-style-react.com/pr_6466

**Comment by shlomitc** [comment]
@sivanharel

**Comment by sivanharel** [comment]
@roibWix - very nice work :)
I have a small question, why do you need the `getDataHook` method in the `input.uni.driver`?

**Comment by roibWix** [comment]
Hey @sivanharel ,
We don't need.
It's my next commit. just wanted to make sure that the unit tests pass

File: src/DatePicker/test/DatePicker.spec.js [hunk 1]
```diff
@@ -1,856 +1,865 @@
...
 
-      it('should hide the focus visually on the current element from the user', () => {
-        const { calendarDriver, driver } = createDriver(
-          <DatePicker onChange={noop} />,
-        );
-        driver.open();
-        expect(calendarDriver.isFocusedDayVisuallyUnfocused()).toBe(true);
-      });
+          await calendarDriver.pressLeftArrow();
+          setTimeout(async () => {
```
Comment (TalZa1810) [inline src/DatePicker/test/DatePicker.spec.js:479]: same comment

File: src/DatePicker/test/DatePicker.spec.js [hunk 2]
```diff
@@ -1,856 +1,865 @@
...
+        expect(await inputDriver.hasRightBorderRadius()).toBe(true);
+        expect(await inputDriver.hasLeftBorderRadius()).toBe(true);
+      });
     });
 
-    describe('when undefined', () => {
-      it('should display `placeholderText`', () => {
-        const { inputDriver } = createDriver(
+    describe('inputProps prop', () => {
+      it('should pass inputProps to input component', async () => {
```
Comment (TalZa1810) [inline src/DatePicker/test/DatePicker.spec.js:826]: After deleting these methods this test can be deleted as well.

File: src/Input/Input.uni.driver.d.ts [hunk 3]
```diff
@@ -57,6 +57,8 @@ export interface InputUniDriver extends BaseUniDriver {
   clickMagnifyingGlass: () => Promise<void>;
   clickMenuArrow: () => Promise<void>;
   hasMenuArrow: () => Promise<boolean>;
+  hasRightBorderRadius: () => Promise<boolean>;
```
Comment (TalZa1810) [inline src/Input/Input.uni.driver.d.ts:60]: delete these methods

File: src/Input/Input.uni.driver.d.ts [hunk 4]
```diff
@@ -57,6 +57,8 @@ export interface InputUniDriver extends BaseUniDriver {
...
+  hasLeftBorderRadius: () => Promise<boolean>;
```
Comment (TalZa1810) [inline src/Input/Input.uni.driver.d.ts:61]: delete

File: src/Input/Input.js [hunk 5]
```diff
@@ -94,7 +94,7 @@ class Input extends Component {
 
   _onFocus = event => {
     const { onFocus } = this.props;
-    this.setState({ focus: true });
+    this._isMounted && this.setState({ focus: true });
```
Comment (TalZa1810) [inline src/Input/Input.js:97]: Why do we need to add this to the logic of the component? Same comment as the DatePicker.js

**Review by TalZa1810 — CHANGES_REQUESTED** [review]
Hi Roi,
I've added some comments. I hope everything is clear. Let me know if something is unclear.

File: src/DatePicker/test/DatePicker.spec.js [hunk 6]
```diff
@@ -1,856 +1,865 @@
...
+  describe('[async]', () => {
+    runTests(createRendererWithUniDriver(datePickerUniDriverFactory));
+  });
 
-    it('should set inputDataHook from props', () => {
-      const { inputDriver } = createDriver(
-        <DatePicker onChange={noop} inputDataHook={'inputDataHook'} />,
-      );
-      expect(inputDriver.getDataHook()).toBe('inputDataHook');
+  async function runTests(render) {
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:31]: done

File: src/DatePicker/test/DatePicker.spec.js [hunk 7]
```diff
@@ -1,856 +1,865 @@
...
+    afterEach(() => cleanup());
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:32]: done

File: src/DatePicker/DatePicker.js [hunk 8]
```diff
@@ -48,6 +48,7 @@ export default class DatePicker extends React.PureComponent {
 
   constructor(props) {
     super(props);
+    this._isMounted = false;
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/DatePicker.js:51]: image

File: src/DatePicker/test/DatePicker.spec.js [hunk 9]
```diff
@@ -1,856 +1,865 @@
...
-      it('should be disabled', () => {
-        const { inputDriver } = createDriver(
-          <DatePicker onChange={noop} disabled />,
+    describe('date picker input', () => {
+      it('should exist', async () => {
+        const {driver} = render(
+          <DatePicker onChange={noop}/>,
         );
-        expect(inputDriver.isDisabled()).toBe(true);
+        expect(await driver.exists()).toBe(true);
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:43]: oopppss,fixed

File: src/DatePicker/test/DatePicker.spec.js [hunk 10]
```diff
@@ -1,856 +1,865 @@
...
+      describe('should close', () => {
+        it('on select date with click', async () => {
+          const {driver: {inputDriver, calendarDriver}} = render(
+            <DatePicker onChange={noop}/>,
+          );
 
-        expect(calendarDriver.isVisible()).toBe(true);
-      });
-    });
+          await inputDriver.trigger('click');
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:129]: done

File: src/DatePicker/test/DatePicker.spec.js [hunk 11]
```diff
@@ -1,856 +1,865 @@
...
 
-        inputDriver.trigger('click');
-        calendarDriver.triggerKeyDown({ key: 'Escape', keyCode: 27 });
+        it('on outside click', async () => {
+          const {driver: {inputDriver, calendarDriver}} = render(
+            <DatePicker onChange={noop}/>,
+          );
 
-        expect(calendarDriver.isVisible()).toBe(false);
+          await inputDriver.trigger('click');
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:181]: done

File: src/Input/Input.js (hunk 5 above)
Comment (roibWix) [reply src/Input/Input.js:97]: image

**Comment by roibWix** [comment]
@TalZa1810  fixed all the simple stuff.

File: src/DatePicker/DatePicker.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/DatePicker.js:51]: talk to moshe

File: src/DatePicker/DatePicker.uni.driver.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  const calendarDriver = calendarUniDriverFactory(
+    findByHook(base, dataHooks.datePickerCalendar),
+    body,
+  );
+  const inputDriver = inputUniDriverFactory(
+    findByHook(base, dataHooks.datePickerInput),
+    body,
+  );
+
+  const driver = {
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/DatePicker.uni.driver.js:16]: add descriptions altoght it doesnt

File: src/DatePicker/README.TESTKIT.md [hunk 13]
```diff
@@ -1,133 +0,0 @@
-# DatePicker TestKits
-
-## DatePicker TestKit API
-
-### Polyfills
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/README.TESTKIT.md:5]: talk with egozi where to place it

File: src/Input/Input.uni.driver.js [hunk 14]
```diff
@@ -98,6 +97,8 @@ export const testkit = (base, body) => {
     getCursorLocation: async () => await input._prop('selectionStart'),
     clearText: () => driver.enterText(''),
     clickOutside: () => ReactBase.clickDocument(),
+    hasRightBorderRadius: async () => !(await base.attr(DATA_ATTR.RIGHTBORDERRADIUS)),
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/Input/Input.uni.driver.js]: deleted

File: src/DatePicker/constants.js [hunk 15]
```diff
@@ -2,4 +2,5 @@ export const dataHooks = {
   datePickerCalendar: 'date-picker-calendar',
   datePickerPopover: 'date-picker-popover',
   datePickerInputContainer: 'date-picker-input-container',
+  datePickerInput: 'date-picker-input',
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/constants.js:5]: done

File: src/Input/Input.uni.driver.js [hunk 16]
```diff
@@ -98,6 +97,8 @@ export const testkit = (base, body) => {
...
+    hasLeftBorderRadius: async () => !(await base.attr(DATA_ATTR.LEFTBORDERRADIUS)),
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/Input/Input.uni.driver.js]: done

File: src/DatePicker/test/DatePicker.spec.js [hunk 17]
```diff
@@ -1,856 +1,865 @@
...
+          const date = new Date(2018, 1, 5);
+          const {driver: {calendarDriver, driver}} = render(
+            <DatePicker onChange={noop} value={date}/>,
+          );
+
+          await driver.open();
+          expect(await calendarDriver.getFocusedDay()).toEqual('5');
+          await calendarDriver.pressLeftArrow();
+          // we need setTimeout because pressLeftArrow trigger async actions
+          setTimeout(async () => {
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:432]: done

File: src/DatePicker/test/DatePicker.spec.js [hunk 18]
```diff
@@ -1,856 +1,865 @@
...
-    });
+          await inputDriver.trigger('click');
+          await calendarDriver.clickOnNthDay();
 
-    describe('should close', () => {
-      it('on select date with click', () => {
-        const { inputDriver, calendarDriver } = createDriver(
-          <DatePicker onChange={noop} />,
-        );
+          setTimeout(async () => expect(await calendarDriver.isVisible()).toBe(false), 0);
```
[the remark this answers is not in the document]
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:132]: done

File: src/DatePicker/test/DatePicker.spec.js (hunk 1 above)
Comment (roibWix) [reply src/DatePicker/test/DatePicker.spec.js:479]: done

File: src/Input/Input.uni.driver.d.ts (hunk 3 above)
Comment (roibWix) [reply src/Input/Input.uni.driver.d.ts]: done

**Review by mosheKerbel — APPROVED** [review]

File: src/DatePicker/README.md [hunk 19]
```diff
@@ -1,8 +0,0 @@
-# DatePicker
```
[the remark this answers is not in the document]
Comment (TalZa1810) [reply src/DatePicker/README.md:1]: I think that the locals description should be added to the `DatePicker customizations` example.

**Review by TalZa1810 — APPROVED** [review]
After merging this PR please create the following PRs:
**First PR:** 
1. Making sure that the uni driver methods are properly documented. I think that the Polyfills which are used within our own tests should be documented as well.
2. Adding the locales description to the `DatePicker customizations` example. 

**Second PR:**
Deleting and moving the relevant Input tests to be visual (the border radius related tests).

**Comment by roibWix** [comment]
Hey @TalZa1810 ,
regarding the first PR:
the polyfills are not relevant. they were created because jsdoc didn't support those methods. They are supported today so i'll remove them.

**Comment by roibWix** [comment]
@TalZa1810 ,
Egozi is now working on the input component so he'll work on the tests.

**Pull request merged**